### DISCUSSION TITLE
Balance Power Sensor

### DISCUSSION THREAD

**Discussion by sanghviharshit** [emporia-vue-local/esphome#29] [Q&A]
Can someone help me write the config for adding a balance sensor? 
I have more than 16 circuits on my panel and Vue 2 only has 16 clamps. So the official app creates a new category as balance power = (total - power from the 16 monitored circuits).

I tried the following code, but it doesn't work well as sometimes it's reporting negative balance power consumption.
```
  - platform: template
    name: "Balance Power"
    lambda: return id(total_power).state
        - id(cir1).state
        - id(cir2).state
        - id(cir3).state
        - id(cir4).state
        - id(cir5).state
        - id(cir6).state
        - id(cir7).state
        - id(cir8).state
        - id(cir9).state
        - id(cir10).state
        - id(cir11).state
        - id(cir12).state
        - id(cir13).state
        - id(cir14).state
        - id(cir15).state
        - id(cir16).state;
    update_interval: 1.44s
    id: balance_power
    unit_of_measurement: "W"
  - platform: total_daily_energy
    name: "Balance Daily Energy"
    power_id: balance_power
    accuracy_decimals: 0
```
![Screen Shot 2022-09-15 at 12 05 48 PM](https://user-images.githubusercontent.com/1029311/190488342-96e747c8-97ac-46ee-9c34-75cd3ca9fddb.png)

**Comment by flaviut** [comment]
That looks right!

Are you perhaps over-counting certain circuits with a `multiply: 2` (or similar) filter?

If you have a 2 of 3 phase system, typical in USA apartments and condos, `multiply: 1.7333` is the correct value. And apparently I've had this set up wrong in my own system for forever :grin:. Split-phase systems should have `multiply: 2`.

**Reply by sanghviharshit** [reply]
I do have a 2 phase system, I didn't know the multiplier was supposed to be 1.7333. Is this documented somewhere?

**Reply by sanghviharshit** [reply]
FWIW, it's a single-family home in the US.

**Reply by flaviut** [reply]
In that case you most likely have a split phase system, and 2 is the correct value. This stuff is confusing, and has to do with how they wired up the transformers. If you have 240V between phases and live in the USA, it's split phase.

Hmm, not sure what's wrong. It seems to be a bit too much to be tolerances stacking up.

Maybe some of your 2-phase loads are drawing power unevenly between phases. When you're having odd readings, for each of your two-phase loads, you can try using a clamp multimeter to see if the power is being drawn evenly between the two wires.

**Reply by sanghviharshit** [reply]
Hmm..interesting. I didn't notice these issues with the readings in Emporia app. I don't have a clamp multimeter, but I will see what I can do. I noticed when the readings is high on one of the 2-phase circuits, it doesn't correlate to the Phase A power and results in a negative balance power.

**Reply by flaviut** [reply]
They might have some sort of smarts in their cloud to either calibrate these sort of problems out or simply hide them :)

Without a clamp multimeter, you could temporary move one of the clamps so that the Vue captures data from both channels of the 2-phase loads.

**Reply by jasonogren** [reply]
Im in the US with split phase and I have a similar sensor defined 'Balance Power' . I have double and triple checked multiplier circuits/phases/etc. I will tell you my specific case findings. When any of my 240V 2 phase circuits are active the number is usually negative. Interestingly when one of my 2 stage Heat Pump/AC is in stage 1 the number is usually positive, stage 2 it is -75 +/- 50 %, when both Heat Pumps are running any combination of Stage 1 and  Stage 2, it varies wildly -150 or greater. When no 2 phase circuits are in use, the number appears to be reasonable and could be accurate but fluctuates +/- 10% with each update. Both heat pumps + other house hold use is ~ 8500 watts with the heat pumps accounting for ~4600 watts.  -150/8500 is 1.7% which is within the 2% tolerance emporia quotes.  However I cant explain why it wouldn't be 2% the other way too. Also, it has me wondering in my case if the heat pump legs are not really balanced so I'm going to find the manual to understand the switching for phase 1 and 2 and  also I will measure both legs of the heat pump circuits, but thinking of carefully cramming all 18 wires back in there is weighing heavy on me..  Of note, I do have a 240v 2 phase subpanel that I measure both legs independently without the multiplier  because they are wildly unbalanced, but I don't know if you have a subpanel. Final note to the developers: this project kicks ass and I couldn't be happier to keep my home assistant cloud free! Thank you so much.

**Reply by sanghviharshit** [reply]
Exactly, this project kicks ass!
I did some digging and I found something interesting - I noticed on one of my 2-phase breakers, the appliance (Microwave/oven) wasn't using 240v so it was uneven. Changing the multiply to 1 seems to have worked for the >500w outliers. The balance sensor still goes below zero sometimes, but I think that's because there is a difference in voltage between phase A and phase B. Do you think updating the multiplier on other 2-phase circuits to account for this difference will fix it?

![Screen Shot 2022-09-22 at 3 55 27 PM](https://user-images.githubusercontent.com/1029311/191865392-9dff7be6-5f3c-4b4b-9cbc-552e0c481da1.png)

**Reply by flaviut** [reply]
> I did some digging and I found something interesting - I noticed on one of my 2-phase breakers, the appliance (Microwave/oven) wasn't using 240v so it was uneven

Either you have a really fancy microwave/oven combo, or someone did a hack job and put a microwave on the oven circuit. If the latter, I'd recommend getting an electrician to take a look.

> because there is a difference in voltage between phase A and phase B

This is one of the shortcomings of the hardware, sometimes the voltages need calibration. See https://github.com/emporia-vue-local/esphome/discussions/21. I'm not confident this will fix anything, but it can't possibly hurt.

> Do you think updating the multiplier on other 2-phase circuits to account for this difference will fix it?

It depends on the loads that are on the circuit. It's not possible to answer without carefully examining them.

**Reply by whoiswes** [reply]
So this might be stupid, but I was having the same issue with occasional negatives (even though I should always have a few hundred watts of balance power).  On a whim, I changed the time from 1.44s to 1s (to match up with the phase total power) and also changed out the lambda to look like this:

`
return id(phase_a_power).state + id(phase_b_power).state
<subtract individual circuits>
`

I don't know why, but now I'm getting a correct balance calculation.

![image](https://user-images.githubusercontent.com/12676280/193424277-177bf1c9-8491-4fc3-bac6-0504b30cbdb8.png)

EDIT: I actually tried reverting back to the original config as a test and it still works.  So I dunno what's up.  But your original snippet seems to be working for me

EDIT2:  After looking at the long term trends, I actually should be near zero balance when the non-monitored circuits are not in use.  So in my case, the occasional negatives I'm seeing are due to the moving averages making things not add up exactly as other circuits fluctuate.  I also have a slight voltage discrepancy and will get that updated later today hopefully.  Will post back again if that helps.

![image](https://user-images.githubusercontent.com/12676280/193461331-8d35b8ed-cc2d-4108-b589-20e6f471777d.png)

**Comment by sanghviharshit** [comment]
@flaviut Looks like the later as the microwave/oven combo is nothing fancy and it clearly says 220V on the label.
> Either you have a really fancy microwave/oven combo, or someone did a hack job and put a microwave on the oven circuit. If the latter, I'd recommend getting an electrician to take a look.

@whoiswes that was very helpful. I think it has something to do with moving averages as well. I changed the update interval to 2.88s for both the total and balance. This seems to have reduced the wildly negative values. I was still seeing up to -12w on the balance sensor where it should be around 60w. I have added this lambda to report 0 instead of negative for now.
```
  - platform: template
    name: "Balance Power"
    lambda: !lambda |-
      if (id(total_power).state
        - id(cir2).state
        - id(cir6).state
        - id(cir12).state
        - id(cir14).state
        - id(cir16).state
        - id(cir20).state
        - id(cir22).state
        - id(cir26).state
        - id(cir23).state
        - id(cir21).state
        - id(cir19).state
        - id(cir17).state
        - id(cir13).state
        - id(cir11).state
        - id(cir9).state
        - id(cir1).state > 0) {
        return id(total_power).state
        - id(cir2).state
        - id(cir6).state
        - id(cir12).state
        - id(cir14).state
        - id(cir16).state
        - id(cir20).state
        - id(cir22).state
        - id(cir26).state
        - id(cir23).state
        - id(cir21).state
        - id(cir19).state
        - id(cir17).state
        - id(cir13).state
        - id(cir11).state
        - id(cir9).state
        - id(cir1).state;
      } else {
        return 0.0;
      }
    update_interval: 2.88s
    id: balance_power
    unit_of_measurement: "W"
```

**Reply by flaviut** [reply]
Having them as a combo is fancy by my standards :)

To be clear—seperate 120V microwave + 240V oven on one circuit → electrical code violation. Single device rated for 240V on one circuit → safe and fine.

Anyway, glad you figured out the timing thing—that's interesting. Not sure why it works though. Possibly some sort of race condition? The MCU only reports new data on a very specific interval, so that's where the 1.44s interval comes from.

**Reply by ScottEgan** [reply]
> ... I think it has something to do with moving averages as well. I changed the update interval to 2.88s for both the total and balance. This seems to have reduced the wildly negative values. I was still seeing up to -12w on the balance sensor where it should be around 60w. I have added this lambda to report 0 instead of negative for now.

I did some investigation and it does seem to be related to the moving averages. I did a quick test where I used the raw unfiltered values to do the balance math and it eliminated my spikes.

![image](https://user-images.githubusercontent.com/6709827/232111018-54831697-062d-498e-988a-8aa212a34c1e.png)

The red line is where I changed the balance math. You can see that before when there were step like changes in power draw the balance calculation would be all over the place. After changing to doing the math and then filtering, the balance spikes have been reduced.

Hope this can help someone.

**Reply by heisenberg2980** [reply]
> I did some investigation and it does seem to be related to the moving averages. I did a quick test where I used the raw unfiltered values to do the balance math and it eliminated my spikes.

@ScottEgan I have the same spikes in the balance entity when high wattage devices turn on, can you share the formula you created for balance to avoid the spikes?

**Comment by MarkoPaasila** [comment]
My solution was to first make a sensor for `Balance Total Daily Energy` and then derive `Balance Power` from that. That was the only way it didn't completely mess up my energy charts in Home-assistant.
